datafusion/core/tests/fifo.rs

```diff
@@ -17,32 +17,208 @@
+        }
+
+        fn table_type(&self) -> TableType {
+            TableType::Temporary
+        }
+
+        async fn scan(
+            &self,
+            _state: &SessionState,
+            projection: Option<&Vec<usize>>,
+            _filters: &[Expr],
+            _limit: Option<usize>,
+        ) -> Result<Arc<dyn ExecutionPlan>> {
+            Ok(Arc::new(StreamingTableExec::try_new(
+                self.0.schema.clone(),
+                vec![Arc::new(FifoRead(self.0.clone())) as _],
+                projection,
+                self.0.sort.clone(),
+                true,
+            )?))
+        }
+
+        async fn insert_into(
+            &self,
+            _state: &SessionState,
+            input: Arc<dyn ExecutionPlan>,
+            _overwrite: bool,
+        ) -> Result<Arc<dyn ExecutionPlan>> {
+            let ordering = match &self.0.sort {
+                Some(order) => Some(order.iter().map(|e| e.clone().into()).collect()),
+                None => None,
+            };
+
+            Ok(Arc::new(FileSinkExec::new(
+                input,
+                Arc::new(FifoWrite(self.0.clone())),
+                self.0.schema.clone(),
+                ordering,
+            )))
+        }
+    }
+
+    struct FifoRead(Arc<FifoConfig>);
+
+    impl PartitionStream for FifoRead {
+        fn schema(&self) -> &SchemaRef {
+            &self.0.schema
+        }
+
+        fn execute(&self, _ctx: Arc<TaskContext>) -> SendableRecordBatchStream {
+            let config = self.0.clone();
+            let schema = self.0.schema.clone();
+            let mut builder = RecordBatchReceiverStreamBuilder::new(schema, 2);
+            let tx = builder.tx();
+            builder.spawn_blocking(move || {
+                let file = File::open(&config.location)?;
+                let reader = ReaderBuilder::new(config.schema.clone()).build(file)?;
+                for b in reader {
+                    if tx.blocking_send(b.map_err(Into::into)).is_err() {
+                        break;
+                    }
+                }
+                Ok(())
+            });
+            builder.build()
+        }
+    }
+
+    #[derive(Debug)]
+    struct FifoWrite(Arc<FifoConfig>);
+
+    impl DisplayAs for FifoWrite {
+        fn fmt_as(&self, _t: DisplayFormatType, f: &mut Formatter) -> std::fmt::Result {
+            write!(f, "{self:?}")
+        }
+    }
+
+    #[async_trait]
+    impl DataSink for FifoWrite {
+        fn as_any(&self) -> &dyn Any {
+            self
+        }
+
+        fn metrics(&self) -> Option<MetricsSet> {
+            None
+        }
+
+        async fn write_all(
+            &self,
+            mut data: SendableRecordBatchStream,
+            _context: &Arc<TaskContext>,
+        ) -> Result<u64> {
+            let config = self.0.clone();
+            let (sender, mut receiver) = tokio::sync::mpsc::channel::<RecordBatch>(2);
+            // Note: FIFO Files support poll so this could use AsyncFd
```

Comment:
This is an example of the potential benefits of a custom TableProvider, it can use external knowledge about the nature of the files in question.